java-util/src/main/java/io/druid/java/util/common/granularity/Granularity.java

```diff
@@ -0,0 +1,538 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.java.util.common.granularity;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import com.google.common.primitives.Longs;
+import io.druid.java.util.common.IAE;
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.joda.time.Interval;
+import org.joda.time.Period;
+import org.joda.time.format.DateTimeFormatter;
+
+import java.util.BitSet;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public abstract class Granularity
+{
+
+  public static final Granularity SECOND = fromString("SECOND");
+  public static final Granularity MINUTE = fromString("MINUTE");
+  public static final Granularity FIVE_MINUTE = fromString("FIVE_MINUTE");
+  public static final Granularity TEN_MINUTE = fromString("TEN_MINUTE");
+  public static final Granularity FIFTEEN_MINUTE = fromString("FIFTEEN_MINUTE");
+  public static final Granularity THIRTY_MINUTE = fromString("THIRTY_MINUTE");
+  public static final Granularity HOUR = fromString("HOUR");
+  public static final Granularity SIX_HOUR = fromString("SIX_HOUR");
+  public static final Granularity DAY = fromString("DAY");
+  public static final Granularity WEEK = fromString("WEEK");
+  public static final Granularity MONTH = fromString("MONTH");
+  public static final Granularity QUARTER = fromString("QUARTER");
+  public static final Granularity YEAR = fromString("YEAR");
+  public static final Granularity ALL = fromString("ALL");
+  public static final Granularity NONE = fromString("NONE");
+
+  static final List<Granularity> PREDEFINED_GRANULARITIES = ImmutableList.of(
+      SECOND, MINUTE, FIVE_MINUTE, TEN_MINUTE, FIFTEEN_MINUTE, THIRTY_MINUTE,
+      HOUR, SIX_HOUR, DAY, WEEK, MONTH, QUARTER, YEAR);
+
+  // Default patterns for parsing paths.
+  final Pattern defaultPathPattern =
+      Pattern.compile(
+          "^.*[Yy]=(\\d{4})/(?:[Mm]=(\\d{2})/(?:[Dd]=(\\d{2})/(?:[Hh]=(\\d{2})/(?:[Mm]=(\\d{2})/(?:[Ss]=(\\d{2})/)?)?)?)?)?.*$"
+      );
+  final Pattern hivePathPattern =
+      Pattern.compile("^.*dt=(\\d{4})(?:-(\\d{2})(?:-(\\d{2})(?:-(\\d{2})(?:-(\\d{2})(?:-(\\d{2})?)?)?)?)?)?/.*$");
+
+  @JsonCreator
+  public static Granularity fromString(String str)
+  {
+    String name = str.toUpperCase();
+    return GranularityType.createGranularity(name);
+  }
+
+  //simple merge strategy on query granularity that checks if all are equal or else
+  //returns null. this can be improved in future but is good enough for most use-cases.
+  public static Granularity mergeGranularities(List<Granularity> toMerge)
+  {
+    if (toMerge == null || toMerge.size() == 0) {
```

Comment:
so do you want to change the behavior? happy to change if its a simple one without many consequences. otherwise I'd prefer to do outside of this pr.